### ISSUE TITLE
Create it.

### ISSUE THREAD

**Issue by balupton** [balupton/api-cache-proxy#1]
/cc @chase

Should solve our problems with rate limiting right? Providing we cache heavily and the user accessing the API still sends the tokens.

**Comment by chase** [comment]
Now we just have to figure out how we are going to cache it.

I'm guessing time-based would be the easiest way, like doing a pull every 12 or 24 hours.

**Comment by balupton** [comment]
I'm thinking just following the etag and expires headers - so do max caching - max caching better than expiring rate limits.

**Comment by chase** [comment]
If I remember correctly, the expiry is set to 1 minute on pages like contributors.
Wait, now that I think about it, why don't we just cache the pages in temporary files like a real browser?

**Comment by balupton** [comment]
So like store them away as a temp file, and use that if rate limit kicks in? That's a good idea.

**Comment by chase** [comment]
I was thinking more along the lines of using the cache to avoid getting the rate limit at all by using the `If-None-Match` header with the ETag for the file that we currently have cached after the first request.
If the response for subsequent requests is `304 Not Modified` we wouldn't have to update the file at all.

![url caching](https://f.cloud.github.com/assets/5411/1187984/b4287148-23ac-11e3-9237-096456f49b60.png)

Basically, it would reduce the load on GitHub's servers while using the caching implementation that they expect. Win-win.

Although if the number of requests by people who don't plugin their GitHub API credentials is limited to 60 in a short time span, having a caching proxy might still be a good idea.

**Comment by balupton** [comment]
Interesting, I've never heard of `If-None-Match` before? Do you have any particularly good resources to read up on it?

**Comment by chase** [comment]
[If-None-Match, straight from the HTTP 1.1 protocol definition by W3](http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.26), is your best resource.

Essentially, it is the most efficient way to cache things as a client using the server's ETag.

Also, Wikipedia has [a nice explanation of the typical usage of ETags](http://en.wikipedia.org/wiki/HTTP_ETag#Typical_usage).

**Comment by balupton** [comment]
Implemented into feedr, need to push all the changes up.
